compiler/rustc_smir/src/rustc_internal/mod.rs

```diff
@@ -163,3 +166,40 @@ pub type Opaque = impl Debug + ToString + Clone;
 pub(crate) fn opaque<T: Debug>(value: &T) -> Opaque {
     format!("{value:?}")
 }
+
+pub struct StableMir {
+    args: Vec<String>,
+    process: fn(TyCtxt<'_>),
+}
+
+impl StableMir {
+    /// Creates a new `StableMir` instance, with given test_function and arguments.
+    pub fn new(args: Vec<String>, process: fn(TyCtxt<'_>)) -> Self {
+        StableMir { args, process }
+    }
+
+    /// Runs the compiler against given target and tests it with `test_function`
+    pub fn run(&mut self) {
```

Comment:
Is there a reason this is split into two separate functions? Also a user may want to pass extra information to the `process` function, so an `FnOnce + Send` closure may be required.

Reply:
Maybe we should change the receiver type to be used by value instead.

Reply:
> Is there a reason this is split into two separate functions? Also a user may want to pass extra information to the `process` function, so an `FnOnce + Send` closure may be required.

I think making it execute on demand gives more freedom to end user, maybe they could store it somewhere execute it elsewhere who knows.

Reply:
> Maybe we should change the receiver type to be used by value instead.

I don't understand, reasoning behind this, wouldn't we have borrow issue if we use it by value ?